modules/xmpp/JingleSessionPC.js

```diff
@@ -663,51 +591,285 @@ JingleSessionPC.prototype.addSource = function (elem) {
+                    XMPPEvents.REMOTE_UFRAG_CHANGED, remoteUfrag);
+        }
+
+        this.peerconnection.setRemoteDescription(
+            remoteDescription,
+            () => {
+                if (this.signalingState === 'closed') {
+                    logger.error("Attempted to call create answer in closed");
+                    return;
+                }
+                this.peerconnection.createAnswer(
+                    (answer) => {
+                        //TODO(brian): should this go elsewhere?
+                        let localUfrag = getUfrag(answer.sdp);
+                        if (localUfrag != this.localUfrag) {
+                            this.localUfrag = localUfrag;
+                            this.room.eventEmitter.emit(
+                                    XMPPEvents.LOCAL_UFRAG_CHANGED, localUfrag);
+                        }
+                        this.peerconnection.setLocalDescription(
+                            answer,
+                            () => { resolve(); },
+                            (error) => { reject("setLocalDescription failed: " + error); }
+                        );
+                    },
+                    (error) => { reject("createAnswer failed: " + error); },
+                    media_constraints
+                );
+            },
+            (error) => { reject("setRemoteDescription failed: " + error); }
+        );
+    });
+};
+
+/**
+ * Replaces oldStream with newStream and performs a single offer/answer
+ *  cycle after both operations are done.  Either oldStream or newStream
+ *  can be null; replacing a valid 'oldStream' with a null 'newStream'
+ *  effectively just removes 'oldStream'
+ * @param oldStream the current stream in use to be replaced
+ * @param newStream the new stream to use
+ * @returns {Promise}
+ */
+JingleSessionPC.prototype.replaceStream = function (oldStream, newStream) {
+    return new Promise((resolve, reject) => {
+        let workFunction = (finishedCallback) => {
+            let oldSdp = new SDP(this.peerconnection.localDescription.sdp);
+            this.removeStreamFromPeerConnection(oldStream);
+            this.addStreamToPeerConnection(newStream);
+            this._renegotiate()
+                .then(() => {
+                    var newSdp = new SDP(this.peerconnection.localDescription.sdp);
+                    this.notifyMySSRCUpdate(oldSdp, newSdp);
+                    resolve();
+                    finishedCallback();
+                }, (errorMsg, error) => {
+                    if (error) {
+                        errorMsg = errorMsg + ": " + error;
+                    }
+                    logger.error(errorMsg);
+                    reject();
+                    finishedCallback();
+                });
+        };
+        this.modificationQueue.push(workFunction);
+    });
+};
+
+/**
+ * Just add the stream to the peerconnection
+ * @param stream either the low-level webrtc MediaStream or 
```

Comment:
a correct way for documenting this would be:
`@param {JitsiLocalTrack|MediaStream} stream either the low-level webrtc MediaStream or...`
this way some IDEs will recognise the type and provide code completion.